compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp

```diff
@@ -212,4 +214,247 @@ LogicalResult setMatmulLoweringConfig(IREE::GPU::TargetAttr target,
+          bool hasIdleThreads =
+              partitionableLoops.size() == 1 && candidate <= subgroupSize;
+          int vectorSize = hasIdleThreads ? 1 : 4;
+          LLVM_DEBUG(llvm::dbgs() << "Use vector size: " << vectorSize << "\n");
+          threadTileSizes[shapeDim] = vectorSize * scaleToByte;
+          workgroupSize[wgDim] = candidate / vectorSize;
+          assert(numThreads % (candidate / vectorSize) == 0);
+          numThreads /= candidate / vectorSize;
+        } else {
+          if (wgDim == 0)
+            vectorizable = false;
+          threadTileSizes[shapeDim] = scaleToByte;
+          workgroupSize[wgDim] = candidate;
+          assert(numThreads % candidate == 0);
+          numThreads /= candidate;
+        }
+        assert(numThreads >= 1);
+        break;
+      }
+
+      // Stop if we have distributed all threads.
+      if (numThreads == 1)
+        break;
+      wgDim++;
+    }
+    return numThreads;
+  };
+
+  // First try to see if we can use up all threads without any loss.
+  if (distributeToThreads(subgroupSize) != 1) {
+    // Otherwise, allow larger and larger loss factor.
+
+    // Threads for distribution. Use 32 at least.
+    int64_t numThreads = std::max(subgroupSize, 32);
+    // We can tolerate (1 / lossFactor) of threads in the workgroup to be idle.
+    int64_t lossFactor = 32;
+
+    for (; lossFactor >= 1; lossFactor >>= 1) {
+      if (distributeToThreads(numThreads, lossFactor) == 1)
+        break;
+    }
+  }
+
+  TileSizesListType tileSizes;
+  tileSizes.push_back(workgroupTileSizes);
+  tileSizes.push_back(threadTileSizes);
+
+  // Attach the MMA schedule as an attribute to the entry point export function
+  // for later access in the pipeline.
+  MLIRContext *context = linalgOp.getContext();
+  SmallVector<NamedAttribute, 1> attrs;
+  Builder b(context);
+  attrs.emplace_back(StringAttr::get(context, "workgroup"),
+                     b.getIndexArrayAttr(workgroupTileSizes));
+
+  attrs.emplace_back(StringAttr::get(context, "thread"),
+                     b.getIndexArrayAttr(threadTileSizes));
+
+  // Heuristic value chosen to limit maximum vector sizes when tiling below.
+  const unsigned maxVectorSize = 32;
+
+  // Try to tile all reductions by some small factor, preferrably 4, when
+  // possible. This gives us a chance to perform vector4 load if an input has
+  // its innnermost dimension being reduction. It also avoids generating too
+  // many instructions when unrolling vector later. We limit the expected
+  // vector size by estimating it from the size of the iteration space tile and
+  // limit it to a reasonable value. We process the loops from inner most to
+  // outer most to try to align loads along inner dimensions.
+  int64_t vectorSize = 1;
+  int64_t numLoops = linalgOp.getNumLoops();
+  SmallVector<utils::IteratorType> iterTypes = linalgOp.getIteratorTypesArray();
+  SmallVector<int64_t> loopTileSizes(numLoops, 0);
+  for (const auto &[reverseIdx, iter] :
```

Comment:
```suggestion
  for (auto [reverseIdx, iter] :
```

This `const &` doesn't do anything